src/main/scala/com/wavesplatform/network/HistoryReplier.scala

```diff
@@ -6,70 +6,69 @@ import com.wavesplatform.settings.SynchronizationSettings
     .maximumSize(historyReplierSettings.maxBlockCacheSize)
     .build(new CacheLoader[ByteStr, Array[Byte]] {
-      override def load(key: ByteStr) = blocking(history.heightOf(key).flatMap(history.blockBytes)).get
+      override def load(key: ByteStr) = history.heightOf(key).flatMap(history.blockBytes).get
     })
 
   override def channelRead(ctx: ChannelHandlerContext, msg: AnyRef): Unit = msg match {
-    case GetSignatures(otherSigs) => Future {
+    case GetSignatures(otherSigs) => Task {
       otherSigs.view
-        .map(parent => parent -> blocking(history.blockIdsAfter(parent, settings.maxChainLength)))
+        .map(parent => parent -> history.blockIdsAfter(parent, settings.maxChainLength))
         .find(_._2.nonEmpty) match {
         case Some((parent, extension)) =>
           log.debug(s"${id(ctx)} Got GetSignatures with ${otherSigs.length}, found common parent $parent and sending ${extension.length} more signatures")
           ctx.writeAndFlush(Signatures(parent +: extension))
-        case None if otherSigs.length == 1 && otherSigs.head == blocking(history.lastBlock).get.uniqueId =>
+        case None if otherSigs.length == 1 && otherSigs.head == history.lastBlock.get.uniqueId =>
           // this is the special case when both nodes only have genesis block
           log.debug(s"${id(ctx)} Both local and remote nodes only have genesis block")
           ctx.writeAndFlush(Signatures(otherSigs))
         case _ =>
           log.debug(s"${id(ctx)} Got GetSignatures with ${otherSigs.length} signatures, but could not find an extension")
       }
-    }
+    }.runAsync
 
-    case GetBlock(sig) =>
-      Future(knownBlocks.get(sig)).foreach { bytes =>
-        ctx.writeAndFlush(RawBytes(BlockMessageSpec.messageCode, bytes))
-      }
+    case GetBlock(sig) => Task {
+      Option(knownBlocks.get(sig)).foreach(bytes =>
+        ctx.writeAndFlush(RawBytes(BlockMessageSpec.messageCode, bytes)))
+    }.runAsync
 
     case mbr@MicroBlockRequest(totalResBlockSig) =>
       log.trace(id(ctx) + "Received " + mbr)
+      Task {
+        Option(knownMicroBlocks.get(totalResBlockSig)).foreach { bytes =>
+          ctx.writeAndFlush(RawBytes(MicroBlockResponseMessageSpec.messageCode, bytes))
+          log.trace(id(ctx) + s"Sent MicroBlockResponse(total=${totalResBlockSig.trim})")
+        }
+      }.runAsync
 
-      Future(knownMicroBlocks.get(totalResBlockSig)).foreach { bytes =>
-        ctx.writeAndFlush(RawBytes(MicroBlockResponseMessageSpec.messageCode, bytes))
-        log.trace(id(ctx) + s"Sent MicroBlockResponse(total=${totalResBlockSig.trim})")
-      }
-
-    case _: Handshake =>
-      Future(blocking(history.score())).onComplete {
-        case Success(score) => ctx.writeAndFlush(LocalScoreChanged(score))
-        case Failure(e) => ctx.fireExceptionCaught(e)
-      }
+    case _: Handshake => Task {
+      ctx.writeAndFlush(LocalScoreChanged(history.score()))
+    }.runAsync
```

Comment:
It could be done without a task.
1. We send a current score
2. We process a block and broadcast a new score.

With Task we may not respond in a given term and could be banned

Reply:
nvm, `history.score` is a potentially long-running operation